### PR TITLE
Refactor listener to queue changes and commit as part of transaction synchronization

### PR DESCRIPTION
**Pull request by longwa**
This change is modeled after the hibernate-envers library, which does essentially the same thing in a native hibernate application. 

Starting with Hibernate 5.2+, a transaction is required for all persistence operations. However, the hibernate listeners can't easily participate in the transaction that is causing the audit events. This poses a problem for ensuring that audit events are only written in the case that a transaction is actually committed.

This change queues the audit events for the entire transaction and essentially flushes them after the transaction has successfully committed. This allows the listener to use its own session and transaction as required by hibernate.

Still need to fix the tests and do some failure/negative testing to see what happens if the audit log commit fails. This change guarantees that no audit log entries will ever be written for a transaction that didn't commit, but it might still be possible to NOT write audit log entries for a transaction that does commit. Might be able to switch to `beforeCommit()` to allow a failure to rollback the outer transaction although that might expose the opposite problem.

### CODE REVIEW

**Comment by longwa** [comment]
Tests need to be reworked since this solution requires the transaction to commit in order to trigger the audit write, but the integration tests use `@Rollback` in order to prevent pollution. In this case, we will have to execute the code under test in a separate transaction and cleanup manually.

File: plugin/src/main/groovy/grails/plugins/orm/auditable/AuditLogTransactionSynchronization.groovy [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+                log.debug("Writing ${pendingAuditInstances.size()} pending audit instances in afterCommit()")
+            }
+            AuditLogListenerUtil.getAuditDomainClass().invokeMethod("withNewTransaction") {
+                for (GormEntity entity in pendingAuditInstances) {
+                    entity.save(failOnError: true)
+                }
+            }
+        }
+        finally {
+            pendingAuditInstances.clear()
```
[the remark this answers is not in the document]
Comment (robertoschwald) [reply plugin/src/main/groovy/grails/plugins/orm/auditable/AuditLogTransactionSynchronization.groovy]: Should be ok like it is.
